accumulator_updater/programs/accumulator_updater/src/lib.rs

```diff
@@ -0,0 +1,269 @@
+mod macros;
+
+use anchor_lang::{
+    prelude::*,
+    solana_program::{
+        program::invoke_signed,
+        system_instruction,
+        sysvar::{
+            self,
+            instructions::get_instruction_relative,
+        },
+    },
+};
+
+declare_id!("Fg6PaFpoGXkYsidMpWTK6W2BeZ7FEfcYkg476zPFsLnS");
+
+#[program]
+pub mod accumulator_updater {
+    use super::*;
+
+    pub fn initialize(ctx: Context<Initialize>) -> Result<()> {
+        let whitelist = &mut ctx.accounts.whitelist;
+        whitelist.bump = *ctx.bumps.get("whitelist").unwrap();
+        Ok(())
+    }
+
+    pub fn add_allowed_program(
+        ctx: Context<AddAllowedProgram>,
+        allowed_program: Pubkey,
+    ) -> Result<()> {
+        let whitelist = &mut ctx.accounts.whitelist;
+        require_keys_neq!(allowed_program, Pubkey::default());
+        require!(
+            !whitelist.allowed_programs.contains(&allowed_program),
+            AccumulatorUpdaterError::DuplicateAllowedProgram
+        );
+        whitelist.allowed_programs.push(allowed_program);
+        Ok(())
+    }
+
+    /// Add new account(s) to be included in the accumulator
+    pub fn add_account<'info>(
+        ctx: Context<'_, '_, '_, 'info, AddAccount<'info>>,
+        base_account: Pubkey,
+        data: Vec<Vec<u8>>,
+        account_type: u32,
+        account_schemas: Vec<u8>,
+    ) -> Result<()> {
+        let cpi_caller = ctx.accounts.whitelist_verifier.is_allowed()?;
+        require_eq!(data.len(), account_schemas.len());
+        let accts = ctx.remaining_accounts;
+        require_eq!(accts.len(), data.len());
+        let mut zip = data.into_iter().zip(account_schemas.into_iter());
+
+        for ai in accts {
+            let (account_data, account_schema) = zip.next().unwrap();
+            let seeds = accumulator_acc_seeds!(cpi_caller, base_account, account_schema);
```

Comment:
macro isn't strictly necessary but in the past i've had some weird lifetime/reference issues arise b/c since the seeds are supposed to be `&[&[&[u8]]]` so but with the macro it avoids a lot of those issues. could just be due to some limitations on the rust borrow checker